guide/token-setup-and-validation.md

```diff
@@ -0,0 +1,351 @@
+
+Below shows code snippets for token generation and token validation.
+
+
+# Default configuration: Opaque tokens
+
+
+## Default token generation
+
+### Authorization server
+```c#
+// Startup.cs
+
+//ConfigureServices
+// The default token format uses ASP.NET Core Data Protection to encrypt and sign tokens.
+// If using ASP.NET Core Identity, this is called for you. It does not hurt if it's called again.
+    services.AddDataProtection();
+
+    //If tokens need to be validated in a separate resource server, configure a shared ASP.NET Core DataProtection with shared key store and application name
+    //See https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-2.1&tabs=aspnetcore2x#setapplicationname
+    services.AddDataProtection()
+        .PersistKeysToFileSystem(new System.IO.DirectoryInfo(@"[UNC PATH]"))
+        .SetApplicationName("[APP NAME]");
+
+// Register the OpenIddict services.
+// Additional configuration is only needed if using Introspection on resource servers
+    services.AddOpenIddict()
+        .AddCore(...)
+
+        .AddServer(options => {
+            //other options left out for brevity
+            //This is required if using introspection on resource servers
+            //This is not needed if resource servers will use shared ASP.NET Core DataProtection
+            //options.EnableIntrospectionEndpoint("/connect/introspect");
+        })
+
+//Configure
+//must come before using MVC
+...
+app.UseAuthentication();
+...
+```
+
+## Default token validation
+
+### Authorization server
+```c#
+// Startup.cs
+
+//ConfigureServices
+services.AddOpenIddict()
+        .AddCore(...)
+
+        .AddServer(...)
+
+        .AddValidation();
+
+//Configure
+//must come before using MVC
+...
+app.UseAuthentication();
+...
+```
+
+### Resource server shared ASP.NET Core DataProtection
+```c#
+// Startup.cs
+
+//ConfigureServices
+//If tokens need to be validated in a separate resource server, configure a shared ASP.NET Core DataProtection with shared key store and application name
+//See https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-2.1&tabs=aspnetcore2x#setapplicationname
+services.AddDataProtection()
+    .PersistKeysToFileSystem(new System.IO.DirectoryInfo(@"[UNC PATH]"))
+    .SetApplicationName("[APP NAME]");
+
+services.AddOpenIddict()
+        //This adds a "Bearer" authentication scheme
+        .AddValidation();
+
+//Optionally set Bearer token authentication as default
+//services.AddAuthentication("Bearer");
+
+//Configure
+//must come before using MVC
+...
+app.UseAuthentication();
+...
+```
+### Resource server - introspection
```

Comment:
In this block, we should also explain that to use introspection, one has to create a new client application and grant it the introspection endpoint permission.